### ISSUE TITLE
[CDC] SIGABRT encountered during stress test

### ISSUE THREAD

**Issue by ajcaldera1** [yugabyte/yugabyte-db#2735]
Jira Link: [DB-23560](https://yugabyte.atlassian.net/browse/DB-23560)
Encountered during stress test in tserver.err file (NO FATAL):
```
pure virtual method called
terminate called without an active exception
*** Aborted at 1572023210 (unix time) try "date -d @1572023210" if you are using GNU date ***
PC: @     0x7f06d40f1817 (unknown)
*** SIGABRT (@0x3e500000d1b) received by PID 3355 (TID 0x7f062bce2700) from PID 3355; stack trace: ***
    @     0x7f06d42ad360 pthread_timedjoin_np
    @     0x7f06d40f1817 (unknown)
    @     0x7f06d40f2b7a (unknown)
    @     0x7f06d466b5cc  __gnu_cxx::__verbose_terminate_handler()
    @     0x7f06d4669636 (unknown)
    @     0x7f06d4669680  std::terminate()
    @     0x7f06d466a12f __cxa_pure_virtual
    @     0x7f06d795aee1 (unknown)
    @     0x7f06d61d1f9c (unknown)
    @     0x7f06d61d21d9 (unknown)
    @     0x7f06d621f9a7  yb::rpc::Connection::HandleTimeout()
    @     0x7f06d5bd4442  yb::(anonymous namespace)::PosixRWFile::Sync()
    @     0x7f06d42a4fc4 (unknown)
    @     0x7f06d41a49bd build_trtable
    @                0x0 (unknown)
```
Backtrace from core file found:
```
#0  0x00007f06d40f1817 in raise () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libc.so.6
#1  0x00007f06d40f2b7a in abort () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libc.so.6
#2  0x00007f06d466b5cd in __gnu_cxx::__verbose_terminate_handler() () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libstdc++.so.6
#3  0x00007f06d4669636 in ?? () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libstdc++.so.6
#4  0x00007f06d4669681 in std::terminate() () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libstdc++.so.6
#5  0x00007f06d466a12f in __cxa_pure_virtual () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libstdc++.so.6
#6  0x00007f06d795aee1 in operator()<yb::Status::OK&, void> (__object=<optimized out>, this=<optimized out>) at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:600
#7  __call<void, 0ul, 1ul> (__args=<optimized out>, this=<optimized out>) at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:1074
#8  operator()<, void> (this=<optimized out>) at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:1133
#9  std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (yb::cdc::CDCReadRpc::*)(yb::Status const&)> (yb::cdc::CDCReadRpc*, yb::Status::OK)> >::_M_invoke(std::_Any_data const&) (__functor=...)
    at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:1871
#10 0x00007f06d61d1f9c in operator() (this=<optimized out>) at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:2267
#11 yb::rpc::OutboundCall::InvokeCallbackSync (this=0x260358d0) at ../../src/yb/rpc/outbound_call.cc:384
#12 0x00007f06d61d21d9 in yb::rpc::InvokeCallbackTask::Run (this=<optimized out>) at ../../src/yb/rpc/outbound_call.cc:165
#13 0x00007f06d621f9a8 in yb::rpc::(anonymous namespace)::Worker::Execute (this=<optimized out>) at ../../src/yb/rpc/thread_pool.cc:99
#14 0x00007f06d5bd4443 in operator() (this=0x1d57ec8) at /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-/Cellar/gcc/5.5.0_4/include/c++/5.5.0/functional:2267
#15 yb::Thread::SuperviseThread (arg=0x1d57ea0) at ../../src/yb/util/thread.cc:741
#16 0x00007f06d42a4fc4 in start_thread () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libpthread.so.0
#17 0x00007f06d41a49bd in clone () from /home/yugabyte/yb-software/yugabyte-2.0.3.0-bbbd34d12fb1255aeee5b61720c8250031d78d84-release-centos-x86_64/linuxbrew/lib/libc.so.6
```

[DB-23560]: https://yugabyte.atlassian.net/browse/DB-23560?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by nspiegelberg** [comment]
> #9  std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (yb::cdc::CDCReadRpc::*)(yb::Status const&)> (yb::cdc::CDCReadRpc*, yb::Status::OK)>

That's the lambda in the new LeaderGetChanges read path in cdc_service.cc

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-2735-analysis.md`):

# Analysis: yugabyte/yugabyte-db#2735 (no Jira cross-link)

**Title:** [CDC] SIGABRT encountered during stress test
**Filed:** 2019-10-26 · **Labels:** area/cdc · **State:** OPEN (as of 2026-08-21)
**Reported version:** 2.0.3.0 (`bbbd34d12fb1255aeee5b61720c8250031d78d84`)
**Failure signature:** `pure virtual method called` / `terminate called without an active exception` → SIGABRT on a tserver rpc thread-pool worker.

## Verdict

NOT-REPRODUCED — the crash is a real 2019 use-after-destruction race on `yb::cdc::CDCReadRpc`, but the racy code existed only in a ~2-week master window (2019-10-23 to 2019-11-07); it was mitigated in `7f917f607e`, further fixed in `bad48536f0` (#3441), and structurally closed in `9741897036`. The class no longer exists on master (renamed `GetChangesRpc`, lifetime rearchitected to shared_ptr retention). Recommended disposition: close as fixed by removal/rearchitecture.

## Root cause

`ent/src/yb/cdc/cdc_rpc.cc` @ `89b9812997` (`CDCReadRpc::SendRpcToTserver`):

```cpp
InvokeAsync(cdc_proxy_.get(),
    PrepareController(),
    std::bind(&CDCReadRpc::Finished, this, Status::OK()));  // raw `this`
```

Mechanism, with citations at commit `89b9812997` ("[#2503] [CDC] Make redirection to tablet leaders for GetChanges asynchronous", landed 2019-10-23 19:11 -1000):

1. `CDCServiceImpl::TabletLeaderGetChanges` (ent/src/yb/cdc/cdc_service.cc @89b98) redirects a follower-received `GetChanges` to the tablet leader via `GetChangesCDCRpc(...)` → `rpc::StartRpc<CDCReadRpc>` (src/yb/rpc/rpc.h:254-259 @89b98). The **only** strong reference to the `CDCReadRpc` is the `rpcs_` registry slot; the completion lambda calls `rpcs_.Unregister(rpc_handle)`, dropping that last reference from *inside* the callback chain.
2. The response callback handed to the outbound RPC layer is `std::bind(&CDCReadRpc::Finished, this, Status::OK())` — a **raw `this`**, holding no reference count. The era code even documents the assumption it violates: `OutboundCall::InvokeCallbackSync` clears `callback_` "since it may be holding onto reference counts via bound parameters" (src/yb/rpc/outbound_call.cc:~384 @89b98) — with raw `this` there is nothing to hold.
3. Under stress, redirected GetChanges calls time out (`yb::rpc::Connection::HandleTimeout` appears in the reporter's tserver.err trace). Timeout and completion paths queue `InvokeCallbackTask` (outbound_call.cc:163-165 @89b98) onto the messenger thread pool. Once any invocation of `Finished` reaches `InvokeCallback` → user lambda → `rpcs_.Unregister`, the object is destroyed while the RPC layer (queued callback task, retrier/timeout path on the reactor) can still reach it through the raw pointer.
4. A late dispatch then calls virtual `Finished` on a destructing object. `RpcCommand::Finished` is pure virtual (src/yb/rpc/rpc.h:71 @89b98; still `= 0` at src/yb/rpc/rpc.h:70 on master), so during base-subobject destruction the vptr points at a vtable whose `Finished` slot is `__cxa_pure_virtual` → "pure virtual method called" → `std::terminate` → SIGABRT.

This matches the reporter's core backtrace frame-for-frame: frame #9 `std::_Bind<std::_Mem_fn<void (yb::cdc::CDCReadRpc::*)(yb::Status const&)>(CDCReadRpc*, Status::OK)>::_M_invoke` ← #11 `OutboundCall::InvokeCallbackSync` (outbound_call.cc:384) ← #12 `InvokeCallbackTask::Run` (outbound_call.cc:165) ← #13 rpc thread_pool worker (thread_pool.cc:99); frames #5-#4 `__cxa_pure_virtual` → `std::terminate`. Contributor nspiegelberg confirmed on-issue (2019-10-27): "That's the lambda in the new LeaderGetChanges read path in cdc_service.cc".

**Version-provenance note:** the reported tag base `bbbd34d` (2019-10-23 10:48 +0300) contains no `CDCReadRpc` at all (`git grep CDCReadRpc bbbd34d` → empty; `bbbd34d` is an ancestor of `89b9812997`). The crash (unix time 1572023210 = 2019-10-25) therefore came from a stress build cut from master hours-to-days after #2503 landed, labeled 2.0.3.0. No shipped 2.0.3.0 release binary contains the crashing class.

### Fix lineage (why master is not affected)

- `7f917f607e` ("[#2365] [#2332] [#2736] [#2733] [#2863] [CDC] Use new GetChanges API on CDC Consumer", 2019-11-07; #2733/#2736 are the same stress-test batch as #2735): added `auto retained = shared_from_this(); // Ensure we don't destruct until after the callback.` at `Finished` entry plus a `called_` idempotence guard in `InvokeCallback` — the direct mitigation, 12 days after the crash. It also added a defensive `resp_.Clear()` in `Finished`.
- `bad48536f0` ("[#3441] [cdc] Avoid race condition with CDCReadRpc", 2020-02-03): removed that `resp_.Clear()` after TSAN confirmed races between CDC's YBClient thread and the TCP reactor thread. Shipped in v2.1 (docs/content/stable/releases/ybdb-releases/end-of-life/v2.1.md:84 — the only remaining `CDCReadRpc` mention in the tree).
- `9741897036` ("[#6169]", 2020-12-29): structural close — the bound callback itself now owns the object: `auto self = std::static_pointer_cast<CDCReadRpc>(shared_from_this()); ... std::bind(&CDCReadRpc::Finished, self, Status::OK())`. The OutboundCall's stored callback keeps the Rpc alive until the callback is destroyed; the dangling-`this` window cannot occur.
- Renames/moves: `60f81b0c67` (#16220, 2023-02-24) moved `ent/src/yb/cdc/` → `src/yb/cdc/`; `9923b950ab` (#16265) renamed `cdc_rpc.cc` → `xcluster_rpc.cc` and `CDCReadRpc` → `GetChangesRpc`.

Master today (`src/yb/cdc/xcluster_rpc.cc`): `GetChangesRpc` at :258; self-retention in `Finished` at :277-278; shared_ptr-bound callback at :337-340. Caller `CDCServiceImpl::TabletLeaderGetChanges` at src/yb/cdc/cdc_service.cc:4116 (register :4123-4143, unregister :4147, `rpcs_.Shutdown()` :4573). Residual observation, not this issue's path: sibling `XClusterWriteRpc` (:209) and `GetCompatibleSchemaVersionRpc` (:112) still bind raw `this`, guarded by `called_` + `~XClusterWriteRpc() { CHECK(called_); }` (:162).

## Trigger matrix

Archaeological (code-inspection) matrix; predictions derived from source before checking history/evidence.

| # | Condition | Predicted | Observed |
|---|-----------|-----------|----------|
| 1 | 2.0.3.0 tag base `bbbd34d` (no `CDCReadRpc`) | Cannot crash in the reported frame — class absent | `git grep CDCReadRpc bbbd34d` → empty; confirmed absent |
| 2 | Master window `89b9812997`..`7f917f607e` (2019-10-23→11-07), GetChanges redirect under stress with timeouts | Raw-`this` callback + single strong ref in `rpcs_` → late dispatch on destructing object → `__cxa_pure_virtual` SIGABRT | Reporter's core backtrace matches frame-for-frame (bind at frame #9, InvokeCallbackSync outbound_call.cc:384, InvokeCallbackTask::Run :165); `Connection::HandleTimeout` present in .err trace |
| 3 | ≥ `7f917f607e` (2.1 era) | Window narrowed by `retained = shared_from_this()` + `called_`; residual `resp_.Clear()` data race remains | YB's testbed hit exactly that residual race → #3441, fixed by `bad48536f0`; v2.1 release note "Avoid race condition with CDCReadRpc" |
| 4 | ≥ `9741897036` (2.4+) | Race structurally closed — callback owns shared_ptr | Code inspection: bind takes `self` (shared_ptr) |
| 5 | master HEAD (2026-08-21) | Not affected; class renamed | `CDCReadRpc` absent from tree; `GetChangesRpc` retains via xcluster_rpc.cc:277-278, :337-340 |

## Reproduction

N/A — explanatory. A live repro is not meaningful or attainable:

- The defective code existed only on master between `89b9812997` (2019-10-23) and `7f917f607e` (2019-11-07). No GA release tag ships that window (`bbbd34d` = 2.0.3.0 predates the class); no official container image reproduces the reporter's binary.
- The failure is a nondeterministic TSAN-class destruction race requiring a multi-node CDC stress load with GetChanges leader-redirects and RPC timeouts; YugabyteDB validated the fix lineage with `twodc-test` (`ybd debug --cxx-test twodc-test -n 20`, per `bad48536f0` test plan) and TSAN.
- On any current build the crashing dispatch cannot occur: the bound callback holds a strong reference (src/yb/cdc/xcluster_rpc.cc:337-340), so the object outlives every queued `InvokeCallbackTask`.

Failure signature to recognize the class of crash (for triage of any recurrence):
`pure virtual method called` + `terminate called without an active exception` with a frame `std::_Bind<std::_Mem_fn<void (yb::...Rpc::*)(yb::Status const&)>...>` under `yb::rpc::OutboundCall::InvokeCallbackSync`.

## Versions tested

- `bbbd34d12fb1255aeee5b61720c8250031d78d84` (2.0.3.0 tag base): class absent (git archaeology).
- `89b9812997` (introduction), `7f917f607e` (mitigation), `bad48536f0` (#3441 fix), `9741897036` (structural fix): diffs inspected.
- master HEAD in this checkout (2026-08-21): **not affected** — `CDCReadRpc` gone; successor `GetChangesRpc` uses shared_ptr retention (src/yb/cdc/xcluster_rpc.cc:258, :277-278, :337-340); `RpcCommand::Finished` still pure virtual at src/yb/rpc/rpc.h:70 but unreachable post-destruction by construction.
- No container repro run (see Reproduction — N/A rationale).

## Proposed fix direction

None required — already fixed upstream in three steps (`7f917f607e`, `bad48536f0`, `9741897036`) and the component was rewritten (xcluster/CDCSDK). Actionable follow-up for maintainers: close #2735 as fixed (duplicate-in-effect of #3441's fix lineage). Optional hardening (separate issue if desired): convert the remaining raw-`this` binds in `XClusterWriteRpc` (src/yb/cdc/xcluster_rpc.cc:209) and `GetCompatibleSchemaVersionRpc` (:112) to the `self`-retaining pattern used by `GetChangesRpc` for uniformity.

## Test strategy

Historical coverage: `twodc-test` (now `src/yb/integration-tests/xcluster/` suites) exercises the leader-redirect GetChanges path; `bad48536f0` extended it to cover the cluster scenario that motivated the removed `Clear`. Any future lifetime change to `src/yb/cdc/xcluster_rpc.cc` should run the xcluster integration tests under TSAN (`./yb_build.sh tsan --cxx-test <xcluster test> -n 20`), which is the sanitizer class that caught #3441.

## Confidentiality check

Artifact reviewed: contains only public GitHub issue content (issue #2735 body/comment), public commit SHAs/messages from yugabyte/yugabyte-db, and source citations from the public repo. No customer identifiers, PII, internal hostnames/URLs, secrets, or internal ticket content. Reporter-provided backtrace paths reference only the public release layout. Safe as-is; note GitHub comments/PRs are prohibited for this task and none were made.
